vecindex: support Cosine and InnerProduct distance metrics

andy-kimball opened the pull request:
#### vecbench: support Cos/IP distance in vecbench SQL provider

Add support for Cosine and InnerProduct distance metrics to the
vecbench SQL provider.

#### vecencoding: update RaBitQ encoding for Cosine/IP

Update the RaBitQ encoding for quantized vectors that use the Cosine
or InnerProduct distance metric to include a CentroidDotProduct
field. This field contains the dot product between the original
vector and the partition centroid. It's used to estimate Cosine and
InnerProduct distances (but is not used for L2 distances).

#### opt: support Cosine and InnerProduct in optimizer

Add support for the Cosine and InnerProduct distance metrics to the
optimizer. This means expanding the GenerateVectorSearch rule to
support other distance operators and updating the vector index
selection logic to ensure that the operator matches the distance
metric on the index.

#### sql: support vector_ip and vector_cosine operator classes

Support additional operator classes for vector index columns. The
vector_ip_op class specifies that the InnerProduct distance metric
should be used, while the vector_cosine class specifies that the
Cosine distance metric should be used. These are alternate ways
to determine the similarity of vectors (the default is to use
Euclidean distance - vector_l2_op).

#### cspann: move vecdist.Metric to vecpb.DistanceMetric

Make vecdist.Metric a protobuf enum so that it can more easily be
included in vecpb.Config. Delete the vecdist package.

#### cspann: move RotAlgorithm to vec.proto

Move the RotAlgorithm definition to vec.proto, since it's part of
the vector index configuration.

andy-kimball committed:
cspann: move RotAlgorithm to vec.proto

Move the RotAlgorithm definition to vec.proto, since it's part of
the vector index configuration.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
cspann: move vecdist.Metric to vecpb.DistanceMetric

Make vecdist.Metric a protobuf enum so that it can more easily be
included in vecpb.Config. Delete the vecdist package.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
opt: support Cosine and InnerProduct in optimizer

Add support for the Cosine and InnerProduct distance metrics to the
optimizer. This means expanding the GenerateVectorSearch rule to
support other distance operators and updating the vector index
selection logic to ensure that the operator matches the distance
metric on the index.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
sql: support vector_ip and vector_cosine operator classes

Support additional operator classes for vector index columns. The
vector_ip_op class specifies that the InnerProduct distance metric
should be used, while the vector_cosine class specifies that the
Cosine distance metric should be used. These are alternate ways
to determine the similarity of vectors (the default is to use
Euclidean distance - vector_l2_op).

Epic: CRDB-42943
Release note: None

andy-kimball committed:
vecencoding: update RaBitQ encoding for Cosine/IP

Update the RaBitQ encoding for quantized vectors that use the Cosine
or InnerProduct distance metric to include a CentroidDotProduct
field. This field contains the dot product between the original
vector and the partition centroid. It's used to estimate Cosine and
InnerProduct distances (but is not used for L2 distances).

Epic: CRDB-42943
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147681)
<!-- Reviewable:end -->

andy-kimball committed:
vecbench: support Cos/IP distance in vecbench SQL provider

Add support for Cosine and InnerProduct distance metrics to the
vecbench SQL provider.

Epic: CRDB-42943
Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

It doesn't have to be in this PR, but before the release we need to add a version check to prevent creating an index with a different opclass until the upgrade is finalized.

> Reviewed 15 of 15 files at r1, 48 of 48 files at r2, 8 of 8 files at r3, 15 of 15 files at r4, 11 of 11 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147681#-:-ORtoW_uAcYcuZ91wjTf:bdo6hln)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`pkg/sql/logictest/testdata/logic_test/vector_index` line 728 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147681#-ORtnkZO1GNmloBVsCha:-ORtnkZO1GNmloBVsChb:b-57hcvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02885f6ee6c28dd48a4d4d4b7a5626a0fd1dc9a/pkg/sql/logictest/testdata/logic_test/vector_index#L728)):*
> ```
> # Results using Cosine distance.
> query ITF rowsort
> SELECT a, v, round(v <=> '[2, 2]', 2) dist FROM distance_metrics ORDER BY v <=> '[2, 2]' LIMIT 5;
> ```

Why only add the hint for inner product?
___
*[`pkg/sql/opt/xform/testdata/rules/limit` line 3739 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147681#-ORtj1ci6f7EJSc7Z2EL:-ORtj1ci6f7EJSc7Z2EM:b-59ja2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/197944f89471c2b1f6b991675a60409a1625d47e/pkg/sql/opt/xform/testdata/rules/limit#L3739)):*
> ```
> 
> # Case with Cosine distance operator.
> opt expect=GenerateVectorSearch
> ```

We should have a test to make sure we don't select a vector index with the wrong metric type.


<!-- Sent from Reviewable.io -->

andy-kimball committed:
vecindex: don't allow Cos/IP op classes to be used until 25.3

Add version check to prevent Cos/IP op classes from being used until
25.3 is finalized. Otherwise, 25.2 binaries will try to use the index
as if it was built with the L2 distance metric.

Epic: CRDB-42943
Release note: None

DrewKimball reviewed (commented):
> Reviewed 75 of 75 files at r7, 48 of 48 files at r8, 8 of 8 files at r9, 15 of 15 files at r10, 2 of 11 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147681)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/opt/xform/testdata/rules/limit` line 3739 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147681#-ORtj1ci6f7EJSc7Z2EL:-ORvlpWT9E_ZGpKKtEWm:b-dw9vo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/830dbae251f129296b954dd85663769b80775724/pkg/sql/opt/xform/testdata/rules/limit#L3739)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Can you give an example of what you're thinking?
</blockquote></details>

For example, force an index that would otherwise apply, but is the wrong metric type. We shouldn't plan a vector search against that index.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
LGTM

andy-kimball commented:
*[`pkg/sql/opt/xform/testdata/rules/limit` line 3739 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147681#-ORtj1ci6f7EJSc7Z2EL:-ORwmM-SEOdTIiE723v_:b-6uu0nu) ([raw file](https://github.com/cockroachdb/cockroach/blob/830dbae251f129296b954dd85663769b80775724/pkg/sql/opt/xform/testdata/rules/limit#L3739)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

For example, force an index that would otherwise apply, but is the wrong metric type. We shouldn't plan a vector search against that index.
</blockquote></details>

There's a logic test that does that already:
```
# Attempt to use index that doesn.t match the distance operator.
statement error index "idx1" cannot be used for this query
SELECT a, v, round(v <=> '[2, 2]', 2) dist FROM distance_metrics@idx1 ORDER BY v <=> '[2, 2]' LIMIT 5;
```


<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r=drewkimball
